src/host/posix/dnssd.cpp

```diff
@@ -0,0 +1,324 @@
+    {
+        mStateChangeCallback(mState);
+    }
+
+exit:
+    return;
+}
+
+Mdns::Publisher::ResultCallback DnssdPlatform::MakePublisherCallback(RequestId aRequestId, RegisterCallback aCallback)
+{
+    return [aRequestId, aCallback](otbrError aError) {
+        if (aCallback != nullptr)
+        {
+            aCallback(aRequestId, OtbrErrorToOtError(aError));
+        }
+    };
+}
+
+void DnssdPlatform::SetDnssdStateChangedCallback(DnssdStateChangeCallback aCallback)
+{
+    mStateChangeCallback = aCallback;
+}
+
+void DnssdPlatform::RegisterService(const Service &aService, RequestId aRequestId, RegisterCallback aCallback)
+{
+    Mdns::Publisher::SubTypeList subTypeList;
+    Mdns::Publisher::TxtData     txtData(aService.mTxtData, aService.mTxtData + aService.mTxtDataLength);
+
+    for (uint16_t index = 0; index < aService.mSubTypeLabelsLength; index++)
+    {
+        subTypeList.push_back(aService.mSubTypeLabels[index]);
+    }
+
+    mPublisher.PublishService(aService.mHostName, aService.mServiceInstance, aService.mServiceType, subTypeList,
+                              aService.mPort, txtData, MakePublisherCallback(aRequestId, aCallback));
+}
+
+void DnssdPlatform::UnregisterService(const Service &aService, RequestId aRequestId, RegisterCallback aCallback)
+{
+    mPublisher.UnpublishService(aService.mServiceInstance, aService.mServiceType,
+                                MakePublisherCallback(aRequestId, aCallback));
+}
+
+void DnssdPlatform::RegisterHost(const Host &aHost, RequestId aRequestId, RegisterCallback aCallback)
+{
+    Mdns::Publisher::AddressList addressList;
+
+    for (uint16_t index = 0; index < aHost.mAddressesLength; index++)
+    {
+        addressList.push_back(Ip6Address(aHost.mAddresses[index].mFields.m8));
+    }
+
+    mPublisher.PublishHost(aHost.mHostName, addressList, MakePublisherCallback(aRequestId, aCallback));
+}
+
+void DnssdPlatform::UnregisterHost(const Host &aHost, RequestId aRequestId, RegisterCallback aCallback)
+{
+    mPublisher.UnpublishHost(aHost.mHostName, MakePublisherCallback(aRequestId, aCallback));
+}
+
+std::string DnssdPlatform::KeyNameFor(const Key &aKey)
+{
+    std::string name(aKey.mName);
+
+    if (aKey.mServiceType != nullptr)
+    {
+        name += ".";
+        name += aKey.mServiceType;
+    }
+
+    return name;
+}
+
+void DnssdPlatform::RegisterKey(const Key &aKey, RequestId aRequestId, RegisterCallback aCallback)
+{
+    Mdns::Publisher::KeyData keyData(aKey.mKeyData, aKey.mKeyData + aKey.mKeyDataLength);
+
+    mPublisher.PublishKey(KeyNameFor(aKey), keyData, MakePublisherCallback(aRequestId, aCallback));
+}
+
+void DnssdPlatform::UnregisterKey(const Key &aKey, RequestId aRequestId, RegisterCallback aCallback)
+{
+    mPublisher.UnpublishKey(KeyNameFor(aKey), MakePublisherCallback(aRequestId, aCallback));
+}
+
+void DnssdPlatform::HandleMdnsState(Mdns::Publisher::State aState)
+{
+    if (mPublisherState != aState)
+    {
+        mPublisherState = aState;
```

Comment:
Do we need to track the `Mdns::Publisher::State` in `mPublisherState`? Is there an API where we can ask the publisher for its current state?

If we have such an API, using that in `UpdateState()` may be better. We would know we are using the most up-to-date state (no matter how `UpdateState()` is called) and we would not need to wait until the `HandleMdnsState()` callback is invoked to use the new publisher state.

Here, in `HandleMdnsState()` we can then simply call `UpdateState()`.

What do you think?

Reply:
That makes sense to me. However there isn't such API for now. And this API will be a virtual method thus need implementation for each sub-class. I'm open to add such API in Mdns if we want to do that. Maybe let's do that in a seperate PR?

Reply:
Sounds good. Thanks. 
I did not know that we don't have an API to ask `Publisher` about its current state. Make sense to add it later.